File: app/src/main/java/com/xephorium/crystalnote/ui/custom/NoteListView.java
package com.xephorium.crystalnote.ui.custom;

/*
  NoteListView.java                                                             04.28.2017
  Christopher Cruzen

    Encapsulating the view behavior of a SwipeRefreshLayout, ScrollView, and two ListViews,
  this class represents a collection of notes organized by reverse chronological order.
  The first ListView is populated by notes modified today and the second is populated by
  notes modified before today.

*/

import android.content.Context;
import android.support.v4.widget.SwipeRefreshLayout;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.TextView;

import com.xephorium.crystalnote.R;
import com.xephorium.crystalnote.data.model.Note;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class NoteListView extends SwipeRefreshLayout {

    @BindView(R.id.new_note_header)
    TextView newNoteHeader;

    @BindView(R.id.new_note_list_view)
    ListView newNoteListView;

    @BindView(R.id.old_note_header)
    TextView oldNoteHeader;

    @BindView(R.id.old_note_list_view)
    ListView oldNoteListView;

    private NoteListViewListener noteListViewListener;

    public NoteListView(Context context) {
        super(context, null);
        buildNoteListView(context);
        ButterKnife.bind(this);
    }

    public NoteListView(Context context, AttributeSet attrs) {
        super(context, attrs);
        buildNoteListView(context);
        ButterKnife.bind(this);
    }

    private void buildNoteListView(Context context) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        View noteListViewLayout = layoutInflater.inflate(R.layout.note_list_view_layout, null);
        this.addView(noteListViewLayout);

        this.setNoteListViewListener(getDefaultNoteListViewListener());
        this.setOnRefreshListener(getOnRefreshListener());
    }

    public void populateNoteList(List<Note> newNotes, List<Note> oldNotes) {
        if (newNotes.size() > 0 && oldNotes.size() > 0) {
            newNoteHeader.setVisibility(View.VISIBLE);
            newNoteListView.setVisibility(View.VISIBLE);
            newNoteListView.setAdapter(getNoteListAdapter(newNotes));
            oldNoteHeader.setVisibility(View.VISIBLE);
            oldNoteListView.setVisibility(View.VISIBLE);
            oldNoteListView.setAdapter(getNoteListAdapter(oldNotes));
        } else if (newNotes.size() > 0) {
            newNoteHeader.setVisibility(View.GONE);
            newNoteListView.setVisibility(View.VISIBLE);
            newNoteListView.setAdapter(getNoteListAdapter(newNotes));
            oldNoteHeader.setVisibility(View.GONE);
            oldNoteListView.setVisibility(View.GONE);
            oldNoteListView.setAdapter(getNoteListAdapter(new ArrayList<Note>()));
        } else if (oldNotes.size() > 0) {
            newNoteHeader.setVisibility(View.GONE);
            newNoteListView.setVisibility(View.GONE);
            newNoteListView.setAdapter(getNoteListAdapter(new ArrayList<Note>()));
            oldNoteHeader.setVisibility(View.GONE);
            oldNoteListView.setVisibility(View.VISIBLE);
            oldNoteListView.setAdapter(getNoteListAdapter(oldNotes));
        } else {
            newNoteHeader.setVisibility(View.GONE);
            newNoteListView.setVisibility(View.GONE);
            newNoteListView.setAdapter(getNoteListAdapter(new ArrayList<Note>()));
            oldNoteHeader.setVisibility(View.GONE);
            oldNoteListView.setVisibility(View.GONE);
            oldNoteListView.setAdapter(getNoteListAdapter(new ArrayList<Note>()));
        }
        newNoteListView.setLayoutParams(getScrollViewHeightParams(newNoteListView));
        oldNoteListView.setLayoutParams(getScrollViewHeightParams(oldNoteListView));
    }

    public void setNoteListViewListener(NoteListViewListener noteListViewListener) {
        this.noteListViewListener = noteListViewListener;
    }

    public NoteListViewListener getNoteListViewListener() {
        return this.noteListViewListener;
    }

    private NoteListViewListener getDefaultNoteListViewListener() {
        return new NoteListViewListener() {
            @Override
            public void onNoteClick(Note note) {
                // Default Behavior; Do Nothing
            }

            @Override
            public void onNoteLongClick(Note note) {
                // Default Behavior; Do Nothing
            }

            @Override
            public void onNoteListRefresh() {
                // Default Behavior; Do Nothing
            }
        };
    }

    private NoteListAdapter getNoteListAdapter(final List<Note> notes) {
        return new NoteListAdapter(notes) {
            @Override
            public View.OnClickListener getOnClickListener(final int position) {
                return new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        noteListViewListener.onNoteClick(notes.get(position));
                    }
                };
            }

            @Override
            public View.OnLongClickListener getOnLongClickListener(final int position) {
                return new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View view) {
                        noteListViewListener.onNoteLongClick(notes.get(position));
                        return true;
                    }
                };
            }
        };
    }

    private SwipeRefreshLayout.OnRefreshListener getOnRefreshListener() {
        return new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                noteListViewListener.onNoteListRefresh();
                setRefreshing(false);
            }
        };
    }

    private ViewGroup.LayoutParams getScrollViewHeightParams(ListView listView) {
        int listViewHeight = 0;
        for (int x = 0; x < listView.getAdapter().getCount(); x++) {
            View listItem = listView.getAdapter().getView(x, null, listView);
            listItem.measure(0, 0);
            listViewHeight += listItem.getMeasuredHeight();
        }

        ViewGroup.LayoutParams params = listView.getLayoutParams();
        params.height = listViewHeight + (listView.getDividerHeight() * (listView.getAdapter().getCount() - 1));
        return params;
    }

    public interface NoteListViewListener {
        void onNoteClick(Note note);
        void onNoteLongClick(Note note);
        void onNoteListRefresh();
    }

}

File: app/src/main/java/com/xephorium/crystalnote/ui/custom/NoteListView.kt
package com.xephorium.crystalnote.ui.custom

import android.content.Context
import android.text.format.DateUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View.OnClickListener
import android.view.View.OnLongClickListener
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.xephorium.crystalnote.data.model.PreviewNote
import com.xephorium.crystalnote.databinding.NoteListLayoutBinding

/*
  NoteListView                                                             05.12.2019
  Christopher Cruzen

    Encapsulating the view behavior of a SwipeRefreshLayout and a custom RecyclerView,
  this class represents a collection of PreviewNotes organized by reverse chronological
  order. The first section is populated by PreviewNotes modified today and the second
  is populated by PreviewNotes modified before today.

*/

class NoteListView : SwipeRefreshLayout {


    /*--- Variable Declarations ---*/

    private lateinit var newNotes: MutableList<PreviewNote>
    private lateinit var oldNotes: MutableList<PreviewNote>

    var noteListViewListener: NoteListViewListener? = null

    private lateinit var binding: NoteListLayoutBinding


    /*--- List View Setup ---*/

    constructor(context: Context) : super(context, null) {
        buildNoteListView(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        buildNoteListView(context)
    }

    private fun buildNoteListView(context: Context) {
        binding = NoteListLayoutBinding.inflate(LayoutInflater.from(context))
        this.addView(binding.root)

        this.noteListViewListener = getDefaultNoteListViewListener()
        this.setOnRefreshListener(getOnRefreshListener())
    }


    /*--- Public Methods ---*/

    fun populateNoteList(noteList: List<PreviewNote>, showArchived: Boolean) {
        parseNotes(noteList)

        binding.listNotes.layoutManager = LinearLayoutManager(context)
        binding.listNotes.adapter = getNoteListAdapter(newNotes, oldNotes, showArchived)
    }


    /*--- Private Utility Methods ---*/

    private fun parseNotes(notes: List<PreviewNote>) {
        val newNotes = mutableListOf<PreviewNote>()
        val oldNotes = mutableListOf<PreviewNote>()

        for (x in notes.indices) {
            if (DateUtils.isToday(notes[x].date.time))
                newNotes.add(notes[x])
            else
                oldNotes.add(notes[x])
        }

        this.newNotes = newNotes
        this.oldNotes = oldNotes
    }

    private fun getNoteListAdapter(
        newNotes: List<PreviewNote>,
        oldNotes: List<PreviewNote>,
        showArchived: Boolean
    ): NoteListAdapter {
        return object : NoteListAdapter(context, newNotes, oldNotes, showArchived) {
            override fun getOnClickListener(note: PreviewNote): OnClickListener {
                return OnClickListener { noteListViewListener?.onNoteClick(note) }
            }

            override fun getOnLongClickListener(note: PreviewNote): OnLongClickListener {
                return OnLongClickListener {
                    noteListViewListener?.onNoteLongClick(note)
                    true
                }
            }
        }
    }

    private fun getDefaultNoteListViewListener() = object : NoteListViewListener {
        override fun onNoteClick(note: PreviewNote) = Unit
        override fun onNoteLongClick(note: PreviewNote) = Unit
        override fun onNoteListRefresh() = Unit
    }

    private fun getOnRefreshListener() = OnRefreshListener {
        noteListViewListener?.onNoteListRefresh()
        isRefreshing = false
    }


    /*--- Action Handling Interface ---*/

    interface NoteListViewListener {
        fun onNoteClick(note: PreviewNote)
        fun onNoteLongClick(note: PreviewNote)
        fun onNoteListRefresh()
    }

}
